src/coreclr/vm/threads.cpp

```diff
@@ -2575,93 +2575,122 @@ void Thread::CoUninitialize()
+            else if (thread->HasThreadState(TS_Finalized))
             {
-                ThreadSuspend::LockThreadStore(ThreadSuspend::SUSPEND_OTHER);
-
-                // We remember the next Thread in the thread store
-                // list before deleting the current one. But we can't
-                // use that Thread pointer now that we release the
-                // thread store lock in the middle of the loop. We
-                // have to start from the beginning of the list every
-                // time. If two threads T1 and T2 race into
-                // CleanupDetachedThreads, then T1 will grab the first
-                // Thread on the list marked for deletion and release
-                // the lock. T2 will grab the second one on the
-                // list. T2 may complete destruction of its Thread,
-                // then T1 might re-acquire the thread store lock and
-                // try to use the next Thread in the thread store. But
-                // T2 just deleted that next Thread.
-                thread = ThreadStore::GetAllThreadList(NULL, 0, 0);
+                STRESS_LOG1(LF_SYNC, LL_INFO1000, "T::CDT - finalized thread 0x%p\n", thread);
+
+                thread->ResetThreadState(TS_Finalized);
+                // We have finalized the managed Thread object.  Now it is time to clean up the unmanaged part
+                thread->DecExternalCount(TRUE);
+                thread = next;
             }
             else
-#endif // DEBUGGING_SUPPORTED
             {
                 thread = next;
             }
         }
-        else if (thread->HasThreadState(TS_Finalized))
-        {
-            STRESS_LOG1(LF_SYNC, LL_INFO1000, "T::CDT - finalized thread 0x%p\n", thread);
 
-            thread->ResetThreadState(TS_Finalized);
-            // We have finalized the managed Thread object.  Now it is time to clean up the unmanaged part
-            thread->DecExternalCount(TRUE);
-            thread = next;
-        }
-        else
-        {
-            thread = next;
-        }
+        s_fCleanFinalizedThread = FALSE;
     }
 
-    s_fCleanFinalizedThread = FALSE;
+    ArrayList::Iterator iter = threadsWithManagedCleanup.Iterate();
+    while (iter.Next())
+    {
+        // During the actual thread shutdown,
+        // it may not be practical for us to run enough managed code to clean up
+        // any managed code that needs to know when a thread exits.
+        // Instead, run that clean up here when the thread object is finalized
+        // (which is definitely after the thread has exited)
+        //
+        // We know its safe to do this work here as we can't have finalized the managed thread object yet
+        // as we're running on the finalizer thread right now.
+        PREPARE_NONVIRTUAL_CALLSITE(METHOD__THREAD__ON_THREAD_EXITING);
+        DECLARE_ARGHOLDER_ARRAY(args, 1);
+        args[ARGNUM_0] = OBJECTREF_TO_ARGHOLDER(((Thread*)iter.GetElement())->GetExposedObject());
```

Comment:
What's keeping the Threads in the temporary list alive?

(The lifetime management of the native Threads is a mess, see e.g. https://github.com/dotnet/runtime/issues/107473.)

Reply:
These managed thread objects can't be GCd until they are finalized, and (from what I can tell) they can't be put on the finalizer queue until after the native thread is marked as detached.

Since we're doing this processing on the finalizer thread after detaching the native threads but before processing any additional objects on the finalizer queue, they can't have been finalized yet. At worst they're currently on the finalizer queue.

Reply:
For the native Thread objects, those are kept alive until after this function runs after the managed thread object has been finalized. As we can't run the managed thread object's finalizer while we're in this function as we're on the finalizer thread, we know that the native thread object hasn't been deleted yet.

Reply:
Revisiting this, looks like the handle to the exposed object could be nulled out while before we get here (as that's a short weak handle and after the cleanup in the detached case they are eligible for GC).

I'll clean this up before this comes out of draft.